modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/recovery/RecoveryFuture.java

```diff
@@ -0,0 +1,503 @@
+
+                            if (msg instanceof TxStateRequest) {
+                                if (log.isDebugEnabled())
+                                    log.debug("Receive tx request [local - " + localNodeConstId + " remote - " + constId + ']');
+
+                                TxStateRequest req = (TxStateRequest)msg;
+
+                                TxStateResponse res = new TxStateResponse();
+
+                                for (GridCacheVersion txVer : req.prepared) {
+                                    boolean preparing = txWalState.isPreparing(txVer);
+                                    boolean rollBacked = txWalState.isRollBacked(txVer);
+
+                                    boolean commited = txWalState.isCommited(txVer);
+                                    boolean prepared = txWalState.isPrepared(txVer);
+
+                                    if (preparing || rollBacked)
+                                        res.rollBackTxs.add(txVer);
+                                    else if (!commited && !prepared) {
+                                        //todo check as not found, need re-scan in past.
+                                        System.err.println("NOT FOUND " + txVer + " localNode " + localNodeConstId);
+
+                                        notFoundTxs.add(txVer);
+
+                                        res.rollBackTxs.add(txVer);
+                                    }
+                                }
+
+                                sendResponse(constId, res);
+                            }
+                            else if (msg instanceof TxStateResponse) {
+                                TxStateResponse res = (TxStateResponse)msg;
+
+                                if (log.isDebugEnabled())
+                                    log.debug("Receive tx response [local - " + localNodeConstId + " remote - " + constId + ']');
+
+                                boolean allReceived;
+
+                                synchronized (this) {
+                                    skipTxs.addAll(res.rollBackTxs);
+
+                                    remaining.remove(constId);
+
+                                    allReceived = remaining.isEmpty();
+                                }
+
+                                if (allReceived)
+                                    onDone(new RecoveryContext(initPtr, skipTxs));
+                            }
+                        }
+                        catch (Throwable e) {
+                            U.error(log, "Fail process recovery message.", e);
+
+                            new Thread(new Runnable() {
+                                @Override public void run() {
+                                    G.stop(true);
+                                }
+                            }).start();
+                        }
+                    }
+                });
+            }
+        });
+
+        recoveryIo.onNodeLeft(new CI1<String>() {
```

Comment:
We usually place onNodeLeft callbacks on futures, this is another indication that RecoveryIo should be removed